### ISSUE TITLE
Attempting to Build/Install For Pop!_OS 22.04 LTS

### ISSUE THREAD

**Issue by thebigG** [lemmy-gtk/lemoa#1]
Hi there,

I followed the instructions on the README:

`sudo apt-get install libgtk-4-dev`
`cargo install --git https://github.com/lemmy-gtk/lemoa.git`

And got the following error:

```
The following warnings were emitted during compilation:




warning: `PKG_CONFIG_ALLOW_SYSTEM_CFLAGS="1" "pkg-config" "--libs" "--cflags" "gobject-2.0" "gobject-2.0 >= 2.74"` did not exit successfully: exit status: 1

error: failed to run custom build command for `gobject-sys v0.17.10`

Caused by:
  process didn't exit successfully: `/tmp/cargo-installZ0Y1Vu/release/build/gobject-sys-be3f058cacff8671/build-script-build` (exit status: 1)
  --- stdout
  cargo:rerun-if-env-changed=GOBJECT_2.0_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR
  cargo:warning=`PKG_CONFIG_ALLOW_SYSTEM_CFLAGS="1" "pkg-config" "--libs" "--cflags" "gobject-2.0" "gobject-2.0 >= 2.74"` did not exit successfully: exit status: 1
  error: could not find system library 'gobject-2.0' required by the 'gobject-sys' crate

  --- stderr
  Requested 'gobject-2.0 >= 2.74' but version of GObject is 2.72.4

warning: build failed, waiting for other jobs to finish...
error: failed to compile `lemoa v0.1.0 (https://github.com/lemmy-gtk/lemoa.git#879ba726)`, intermediate artifacts can be found at `/tmp/cargo-installZ0Y1Vu`



```

Here's some info my about system and rust environment:

```
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os

```

```

rustc 1.72.0-nightly (3b2073f07 2023-06-17)

```

Any ideas on what I'm missing? I'm sure it's something obvious and silly...

Thanks!

**Comment by Bnyro** [comment]
Hmm, this looks pretty much like Ubuntu/PopOs ships a very outdated version of the package `gobject-2.0`. You can either try to run
```
git clone https://github.com/lemmy-gtk/lemoa
```
in Cargo.toml, change
```relm4 = "0.6.0" ``` to 0.5.0 and same for ```relm4-components```
and then
```cargon run```.
The other option would be that we try to add a CI to the repo so that there are automated builds for binaries, that can be downloaded and just work.

**Comment by thebigG** [comment]
> The other option would be that we try to add a CI to the repo so that there are automated builds for binaries, that can be downloaded and just work.

I could probably even add something like a Dockerfile so that it is reproducible exactly as it is in CI.

**Comment by thebigG** [comment]
> Hmm, this looks pretty much like Ubuntu/PopOs ships a very outdated version of the package `gobject-2.0`. You can either try to run
> 
> ```
> git clone https://github.com/lemmy-gtk/lemoa
> ```
> 
> in Cargo.toml, change `relm4 = "0.6.0" ` to 0.5.0 and same for `relm4-components` and then `cargon run`. The other option would be that we try to add a CI to the repo so that there are automated builds for binaries, that can be downloaded and just work.

Thanks a lot for this! It helped!

However I came across another issue after going back to previous versions of relm deps:
```
-linux-gnu/lib" "-o" "/home/lgomez/lemoa/target/debug/build/syn-e846c4791d691919/build_script_build-e846c4791d691919" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-fuse-ld=mold"
  = note: collect2: fatal error: cannot find 'ld'
```


I noticed it looks like this project is using the mold linker.

Then I installed the mold linker like this:

```
sudo apt-get install mold
```


Now I don't get the error above anymore(since it is able to find the mold linker on my system), however, I do get this now from rustc:

```
error[E0407]: method `forward_to_parent` is not a member of trait `FactoryComponent`
   --> src/components/post_row.rs:105:5
    |
105 |     fn forward_to_parent(output: Self::Output) -> Option<Self::ParentInput> { Some(output) }
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a member of trait `FactoryComponent`

error[E0407]: method `forward_to_parent` is not a member of trait `FactoryComponent`
  --> src/components/community_row.rs:73:5
   |
73 | /     fn forward_to_parent(output: Self::Output) -> Option<Self::ParentInput> {
74 | |         Some(output)
75 | |     }
   | |_____^ not a member of trait `FactoryComponent`

error[E0407]: method `forward_to_parent` is not a member of trait `FactoryComponent`
  --> src/components/comment_row.rs:72:5
   |
72 | /     fn forward_to_parent(output: Self::Output) -> Option<Self::ParentInput> {
73 | |         Some(output)
74 | |     }
   | |_____^ not a member of trait `FactoryComponent`

error[E0220]: associated type `Index` not found for `Self`
   --> src/components/post_row.rs:116:28
    |
116 |             _index: &Self::Index,
    |                            ^^^^^ associated type `Index` not found

error[E0220]: associated type `Index` not found for `Self`
  --> src/components/community_row.rs:85:28
   |
85 |             _index: &Self::Index,
   |                            ^^^^^ associated type `Index` not found

error[E0220]: associated type `Index` not found for `Self`
  --> src/components/comment_row.rs:84:28
   |
84 |             _index: &Self::Index,
   |                            ^^^^^ associated type `Index` not found

Some errors have detailed explanations: E0220, E0407.
For more information about an error, try `rustc --explain E0220`.
error: could not compile `lemoa` (bin "lemoa") due to 6 previous errors

```

Maybe I'm not using the right rust compiler(?)

Here's the rust version I'm using: `rustc 1.72.0-nightly (3b2073f07 2023-06-17)`

**Comment by Bnyro** [comment]
Hmm, unfortunately it seems like we use a feature of Relm4 that's only available in 0.6.0, I didn't think that would be the case :/

And sorry for the mold thing, I only wanted to use that to speed build times up during development, I forgot to add the .cargo folder to the gitignore file so that it's not a dependency when building normally.

I guess the only solutions we have is that you either try to manually update gobject-2.0 (probably a bad idea) or I provide some working builds using CI.
